File: src/score.py
"""
路径评分与排序模块（Phase 3）。

对匹配后的换乘方案进行多维度评分，按人性化偏好排序。
V1：Sigmoid 平滑加权评分函数。
"""

import math
from dataclasses import dataclass

from matcher import MatchedRoute


def sigmoid(x: float, k: float = 1.0, midpoint: float = 0.0) -> float:
    """Sigmoid 平滑函数：1 / (1 + exp(-k * (x - midpoint)))。"""
    try:
        return 1.0 / (1.0 + math.exp(-k * (x - midpoint)))
    except OverflowError:
        return 0.0 if k * (x - midpoint) < 0 else 1.0


@dataclass
class ScoredRoute:
    """评分后的路径方案。"""
    route: MatchedRoute
    score: float
    scores_detail: dict[str, float]  # 各维度得分明细


def score_routes(routes: list[MatchedRoute]) -> list[ScoredRoute]:
    """
    对路径方案评分并排序（分数越高越好）。

    评分维度：
    1. 总耗时（惩罚长耗时）
    2. 换乘次数（惩罚换乘）
    3. 凌晨时段（惩罚 23:00-06:00 的乘车或换乘）
    4. 里程效率（惩罚绕路）

    权重可调，当前为初始经验值。
    """
    if not routes:
        return []

    # 归一化参考值
    max_time = max(r.total_minutes for r in routes) or 1
    max_dist = max(r.total_distance for r in routes) or 1
    max_transfers = max(len(r.transfer_stations) for r in routes) or 1

    scored = []
    for route in routes:
        details = {}

        # 1. 总耗时得分（越短越高，Sigmoid 平滑）
        time_ratio = route.total_minutes / max_time
        details["耗时"] = 1.0 - sigmoid(time_ratio, k=4.0, midpoint=0.5)

        # 2. 换乘次数得分（越少越高）
        transfers = len(route.transfer_stations)
        details["换乘"] = 1.0 - sigmoid(transfers, k=2.0, midpoint=1.0)

        # 3. 凌晨惩罚（检查每段的发车/到达时间）
        midnight_penalty = _midnight_penalty(route)
        details["凌晨"] = 1.0 - midnight_penalty

        # 4. 里程效率（越短越高）
        dist_ratio = route.total_distance / max_dist
        details["里程"] = 1.0 - sigmoid(dist_ratio, k=3.0, midpoint=0.5)

        # 加权总分
        weights = {
            "耗时": 0.40,
            "换乘": 0.30,
            "凌晨": 0.15,
            "里程": 0.15,
        }
        total = sum(details[k] * weights[k] for k in weights)

        scored.append(ScoredRoute(route=route, score=total, scores_detail=details))

    # 按分数降序
    scored.sort(key=lambda s: s.score, reverse=True)
    return scored


def _midnight_penalty(route: MatchedRoute) -> float:
    """计算凌晨时段惩罚（0~1，越高越差）。"""
    bonus_count = 0
    for seg in route.segments:
        # 发车时间
        if seg.depart_time:
            h = int(seg.depart_time.split(":")[0])
            if 23 <= h or h < 6:
                bonus_count += 1
        # 到达时间
        if seg.arrive_time:
            h = int(seg.arrive_time.split(":")[0])
            if 23 <= h or h < 6:
                bonus_count += 1

    # 换乘等待中跨凌晨
    # （简化：检查换乘等待时间是否包含凌晨时段）
    for i in range(len(route.segments) - 1):
        arrive = route.segments[i].arrive_time
        depart = route.segments[i + 1].depart_time
        if arrive and depart:
            ah = int(arrive.split(":")[0])
            dh = int(depart.split(":")[0])
            # 如果到达在深夜，发车在凌晨后
            if ah >= 22 and (dh < 6 or dh < ah):
                bonus_count += 1

    if bonus_count == 0:
        return 0.0
    # 使用平滑函数
    return sigmoid(bonus_count, k=1.5, midpoint=1.0)


def format_route(scored: ScoredRoute) -> str:
    """格式化输出一条路径。"""
    r = scored.route
    lines = []
    lines.append(f"  {'─' * 60}")
    lines.append(f"  得分: {scored.score:.3f}  "
                 f"| 总耗时: {r.total_minutes // 60}h{r.total_minutes % 60:02d}m  "
                 f"| 换乘: {len(r.transfer_stations)}次  "
                 f"| 里程: {r.total_distance}km")
    lines.append(f"  详细: 耗时={scored.scores_detail['耗时']:.2f}  "
                 f"换乘={scored.scores_detail['换乘']:.2f}  "
                 f"凌晨={scored.scores_detail['凌晨']:.2f}  "
                 f"里程={scored.scores_detail['里程']:.2f}")

    for i, seg in enumerate(r.segments):
        lines.append(f"  [{i + 1}] {seg.train_code}  "
                     f"{seg.from_station} {seg.depart_time} → "
                     f"{seg.to_station} {seg.arrive_time}  "
                     f"({seg.travel_minutes}分钟, {seg.distance}km)")

        # 如果有换乘，在段之间标注
        if i < len(r.segments) - 1:
            lines.append(f"       ╚ 换乘 @" + r.transfer_stations[i] if i < len(r.transfer_stations) else " ╚ 换乘")

    return "\n".join(lines)

File: rust/src/score.rs
//! 路径评分（对齐 Python src/main.py::score_routes 六维评分 + 降序稳定排序）。

use crate::models::{PathSegment, RouteResult};
use std::cmp::Ordering;

/// 对路线按综合分降序评分（分数越高越好；同分保持原序）。
/// 返回 (score, 原始索引) 列表。
pub fn score_routes(routes: &[RouteResult]) -> Vec<(f64, usize)> {
    if routes.is_empty() {
        return Vec::new();
    }
    let max_t = routes.iter().map(|r| r.total_minutes).max().unwrap().max(1) as f64;
    let max_d = routes.iter().map(|r| r.rail_distance).max().unwrap().max(1) as f64;
    let mut scored: Vec<(f64, usize)> = Vec::with_capacity(routes.len());
    for (idx, r) in routes.iter().enumerate() {
        // 凌晨时段惩罚：23:00-06:00 的乘车/换乘（对齐 Python 逐段逐时刻判断）
        let mut night_penalty = 0f64;
        for seg in &r.segments {
            match seg {
                PathSegment::Train(s) => {
                    for m in [s.depart_minutes, s.arrive_minutes] {
                        let h = m.rem_euclid(1440) / 60;
                        if h >= 23 || h < 6 {
                            night_penalty += 1.0;
                        }
                    }
                }
                PathSegment::Interstation(s) => {
                    for m in [s.start_minutes, s.end_minutes] {
                        let h = m.rem_euclid(1440) / 60;
                        if h >= 23 || h < 6 {
                            night_penalty += 1.0;
                        }
                    }
                }
            }
        }
        let inter_penalty = r.interstation_minutes as f64 / 60.0f64.max(1.0);
        let s = 0.0f64.max(1.0 - r.total_minutes as f64 / max_t) * 0.35
            + 0.0f64.max(1.0 - (r.train_transfers as f64 + r.interstation_transfers as f64 * 0.5) / 4.0) * 0.25
            + 0.0f64.max(1.0 - night_penalty * 0.1) * 0.10
            + 0.0f64.max(1.0 - r.rail_distance as f64 / max_d) * 0.15
            + 0.0f64.max(1.0 - inter_penalty / 300.0) * 0.15;
        scored.push((s, idx));
    }
    // 稳定降序（Python sort 稳定 + key=-s）
    scored.sort_by(|a, b| b.0.partial_cmp(&a.0).unwrap_or(Ordering::Equal));
    scored
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::{PathSegment, TrainSegment};

    fn route(
        total_minutes: i32,
        train_xfers: usize,
        inter_xfers: usize,
        rail_distance: i32,
        inter_minutes: i32,
        depart: i32,
        arrive: i32,
    ) -> RouteResult {
        RouteResult {
            segments: vec![PathSegment::Train(TrainSegment {
                train_code: "G1".into(),
                from_station: "北京南".into(),
                to_station: "上海虹桥".into(),
                depart_minutes: depart,
                arrive_minutes: arrive,
                travel_minutes: arrive - depart,
                distance: rail_distance,
            })],
            actual_origin: "北京南".into(),
            actual_destination: "上海虹桥".into(),
            first_departure: depart,
            final_arrival: arrive,
            total_minutes,
            rail_distance,
            train_transfers: train_xfers,
            interstation_transfers: inter_xfers,
            interstation_minutes: inter_minutes,
            transfer_cities: Vec::new(),
            matched_transfer_constraint: false,
        }
    }

    #[test]
    fn empty_input() {
        assert!(score_routes(&[]).is_empty());
    }

    #[test]
    fn faster_route_scores_higher() {
        // 白天、无换乘、同距离：4h vs 8h
        let fast = route(240, 0, 0, 1200, 0, 480, 720);
        let slow = route(480, 0, 0, 1200, 0, 480, 960);
        let scored = score_routes(&[fast, slow]);
        assert!(scored[0].0 > scored[1].0);
    }

    #[test]
    fn night_travel_penalized() {
        let day = route(300, 0, 0, 1000, 0, 480, 780); // 08:00-13:00
        let night = route(300, 0, 0, 1000, 0, 1380, 240); // 23:00-04:00(+1) 全夜间
        let scored = score_routes(&[day, night]);
        assert!(scored[0].0 > scored[1].0);
    }

    #[test]
    fn fewer_transfers_scores_higher() {
        let direct = route(300, 0, 0, 1000, 0, 480, 780);
        let xfer = route(300, 1, 0, 1000, 0, 480, 780);
        let scored = score_routes(&[direct, xfer]);
        assert!(scored[0].0 > scored[1].0);
    }

    #[test]
    fn stable_order_for_ties() {
        let a = route(300, 0, 0, 1000, 0, 480, 780);
        let b = route(300, 0, 0, 1000, 0, 480, 780);
        let c = route(300, 0, 0, 1000, 0, 480, 780);
        let scored = score_routes(&[a, b, c]);
        let idxs: Vec<usize> = scored.iter().map(|(_, i)| *i).collect();
        assert_eq!(idxs, vec![0, 1, 2]); // 同分保持原序
    }

    #[test]
    fn score_in_zero_one_range() {
        let routes = vec![
            route(200, 0, 0, 800, 0, 480, 680),
            route(600, 3, 1, 2000, 120, 300, 900),
        ];
        for (s, _) in score_routes(&routes) {
            assert!((0.0..=1.0).contains(&s));
        }
    }
}
